File: backend-fastapi/app/routers/groups.py
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy import text

from ..auth import current_user, now_iso
from ..db import get_conn
from ..schemas import Balance, GroupCreate, GroupDetail, GroupOut, GroupPatch, MemberAdd, SettleRequest, SettlementOut

router = APIRouter(prefix="/groups")


def is_member(conn, group_id, user_id):
    return conn.execute(
        text("SELECT 1 FROM group_members WHERE group_id = :g AND user_id = :u"), {"g": group_id, "u": user_id}
    ).first() is not None


def load_group(conn, group_id):
    row = conn.execute(text("SELECT id, name, created_by FROM groups WHERE id = :id"), {"id": group_id}).mappings().first()
    if not row:
        raise HTTPException(404, "group not found")
    return dict(row)


def group_detail(conn, group):
    members = conn.execute(
        text(
            "SELECT u.id, u.name, u.email FROM users u JOIN group_members m ON m.user_id = u.id "
            "WHERE m.group_id = :g ORDER BY u.id"
        ),
        {"g": group["id"]},
    ).mappings().all()
    return {**group, "members": [dict(m) for m in members]}


@router.get("", response_model=list[GroupOut])
def list_groups(user=Depends(current_user), conn=Depends(get_conn)):
    rows = conn.execute(
        text(
            "SELECT g.id, g.name, g.created_by FROM groups g JOIN group_members m ON m.group_id = g.id "
            "WHERE m.user_id = :u ORDER BY g.id"
        ),
        {"u": user["id"]},
    ).mappings().all()
    return [dict(r) for r in rows]


@router.post("", response_model=GroupOut, status_code=201)
def create_group(body: GroupCreate, user=Depends(current_user), conn=Depends(get_conn)):
    result = conn.execute(
        text("INSERT INTO groups (name, created_by, created_at) VALUES (:name, :by, :now)"),
        {"name": body.name, "by": user["id"], "now": now_iso()},
    )
    group_id = result.lastrowid
    conn.execute(text("INSERT INTO group_members (group_id, user_id) VALUES (:g, :u)"), {"g": group_id, "u": user["id"]})
    return {"id": group_id, "name": body.name, "created_by": user["id"]}


@router.get("/{id}", response_model=GroupDetail)
def get_group(id: int, user=Depends(current_user), conn=Depends(get_conn)):
    return group_detail(conn, load_group(conn, id))


@router.patch("/{id}", response_model=GroupOut)
def patch_group(id: int, body: GroupPatch, user=Depends(current_user), conn=Depends(get_conn)):
    group = load_group(conn, id)
    if body.name is not None:
        conn.execute(text("UPDATE groups SET name = :name WHERE id = :id"), {"name": body.name, "id": id})
        group["name"] = body.name
    return group


@router.delete("/{id}", status_code=204, response_class=Response)
def delete_group(id: int, user=Depends(current_user), conn=Depends(get_conn)):
    group = load_group(conn, id)
    if group["created_by"] != user["id"]:
        raise HTTPException(403, "only the creator can delete a group")
    p = {"g": id}
    conn.execute(text("DELETE FROM expense_splits WHERE expense_id IN (SELECT id FROM expenses WHERE group_id = :g)"), p)
    conn.execute(text("DELETE FROM expenses WHERE group_id = :g"), p)
    conn.execute(text("DELETE FROM settlements WHERE group_id = :g"), p)
    conn.execute(text("DELETE FROM group_members WHERE group_id = :g"), p)
    conn.execute(text("DELETE FROM groups WHERE id = :g"), p)


@router.post("/{id}/members", response_model=GroupDetail)
def add_member(id: int, body: MemberAdd, user=Depends(current_user), conn=Depends(get_conn)):
    group = load_group(conn, id)
    if not is_member(conn, id, user["id"]):
        raise HTTPException(404, "group not found")
    target = conn.execute(text("SELECT id FROM users WHERE email = :email"), {"email": body.email}).first()
    if not target:
        raise HTTPException(404, "user not found")
    if not is_member(conn, id, target[0]):
        conn.execute(text("INSERT INTO group_members (group_id, user_id) VALUES (:g, :u)"), {"g": id, "u": target[0]})
    return group_detail(conn, group)


@router.get("/{id}/balances", response_model=list[Balance])
def balances(id: int, user=Depends(current_user), conn=Depends(get_conn)):
    load_group(conn, id)
    rows = conn.execute(
        text(
            """
            SELECT u.id AS user_id, u.name,
              COALESCE((SELECT SUM(amount_paise) FROM expenses WHERE group_id = :g AND paid_by = u.id), 0)
            - COALESCE((SELECT SUM(s.share_paise) FROM expense_splits s JOIN expenses e ON e.id = s.expense_id
                        WHERE e.group_id = :g AND s.user_id = u.id), 0)
            + COALESCE((SELECT SUM(amount_paise) FROM settlements WHERE group_id = :g AND from_user = u.id), 0)
            - COALESCE((SELECT SUM(amount_paise) FROM settlements WHERE group_id = :g AND to_user = u.id), 0)
              AS net_paise
            FROM users u JOIN group_members m ON m.user_id = u.id
            WHERE m.group_id = :g ORDER BY u.id
            """
        ),
        {"g": id},
    ).mappings().all()
    return [dict(r) for r in rows]


@router.post("/{id}/settle", response_model=SettlementOut, status_code=201)
def settle(id: int, body: SettleRequest, user=Depends(current_user), conn=Depends(get_conn)):
    load_group(conn, id)
    if not is_member(conn, id, user["id"]):
        raise HTTPException(404, "group not found")
    if body.from_user == body.to_user:
        raise HTTPException(400, "from_user and to_user must differ")
    if not (is_member(conn, id, body.from_user) and is_member(conn, id, body.to_user)):
        raise HTTPException(400, "both users must be group members")
    created_at = now_iso()
    result = conn.execute(
        text(
            "INSERT INTO settlements (group_id, from_user, to_user, amount_paise, created_at) "
            "VALUES (:g, :f, :t, :a, :now)"
        ),
        {"g": id, "f": body.from_user, "t": body.to_user, "a": body.amount_paise, "now": created_at},
    )
    return {"id": result.lastrowid, "group_id": id, **body.model_dump(), "created_at": created_at}

File: backend-node/src/routes/groups.ts
import { Router } from 'express';
import { z } from 'zod';
import type { Db } from '../db.js';
import type { User } from '../auth.js';
import { HttpError, idParam, parseBody } from '../http.js';

type Group = { id: number; name: string; created_by: number };

const nameBody = z.object({ name: z.string().min(1) });
const memberBody = z.object({ email: z.email() });
const settleBody = z.object({ from_user: z.number().int(), to_user: z.number().int(), amount_paise: z.number().int().min(1) });

export function isMember(db: Db, groupId: number, userId: number): boolean {
  return !!db.prepare('SELECT 1 FROM group_members WHERE group_id = ? AND user_id = ?').get(groupId, userId);
}

export function memberIds(db: Db, groupId: number): number[] {
  return (db.prepare('SELECT user_id FROM group_members WHERE group_id = ? ORDER BY user_id').all(groupId) as { user_id: number }[]).map(
    (r) => r.user_id,
  );
}

export function requireGroup(db: Db, groupId: number): Group {
  const group = db.prepare('SELECT id, name, created_by FROM groups WHERE id = ?').get(groupId) as Group | undefined;
  if (!group) throw new HttpError(404, 'group not found');
  return group;
}

export function requireMembership(db: Db, groupId: number, userId: number): Group {
  const group = requireGroup(db, groupId);
  if (!isMember(db, groupId, userId)) throw new HttpError(404, 'group not found');
  return group;
}

function groupDetail(db: Db, group: Group) {
  const members = db
    .prepare('SELECT u.id, u.name, u.email FROM group_members gm JOIN users u ON u.id = gm.user_id WHERE gm.group_id = ? ORDER BY u.id')
    .all(group.id) as User[];
  return { ...group, members };
}

export function groupsRouter(db: Db) {
  const r = Router();

  r.get('/groups', (req, res) => {
    const rows = db
      .prepare('SELECT g.id, g.name, g.created_by FROM groups g JOIN group_members gm ON gm.group_id = g.id WHERE gm.user_id = ? ORDER BY g.id')
      .all(req.user.id);
    res.json(rows);
  });

  r.post('/groups', (req, res) => {
    const body = parseBody(nameBody, req.body);
    const group = db.transaction(() => {
      const result = db
        .prepare('INSERT INTO groups (name, created_by, created_at) VALUES (?, ?, ?)')
        .run(body.name, req.user.id, new Date().toISOString());
      const id = Number(result.lastInsertRowid);
      db.prepare('INSERT INTO group_members (group_id, user_id) VALUES (?, ?)').run(id, req.user.id);
      return { id, name: body.name, created_by: req.user.id };
    })();
    res.status(201).json(group);
  });

  r.get('/groups/:id', (req, res) => {
    const group = requireGroup(db, idParam(req));
    res.json(groupDetail(db, group));
  });

  r.patch('/groups/:id', (req, res) => {
    const group = requireGroup(db, idParam(req));
    const body = parseBody(nameBody.partial(), req.body);
    if (body.name !== undefined) {
      db.prepare('UPDATE groups SET name = ? WHERE id = ?').run(body.name, group.id);
      group.name = body.name;
    }
    res.json(group);
  });

  r.delete('/groups/:id', (req, res) => {
    const group = requireMembership(db, idParam(req), req.user.id);
    if (group.created_by !== req.user.id) throw new HttpError(403, 'only the group creator can delete it');
    db.prepare('DELETE FROM groups WHERE id = ?').run(group.id);
    res.status(204).end();
  });

  r.post('/groups/:id/members', (req, res) => {
    const group = requireMembership(db, idParam(req), req.user.id);
    const body = parseBody(memberBody, req.body);
    const user = db.prepare('SELECT id FROM users WHERE email = ?').get(body.email.toLowerCase()) as { id: number } | undefined;
    if (!user) throw new HttpError(404, 'no user with that email');
    db.prepare('INSERT OR IGNORE INTO group_members (group_id, user_id) VALUES (?, ?)').run(group.id, user.id);
    res.json(groupDetail(db, group));
  });

  r.get('/groups/:id/balances', (req, res) => {
    const group = requireGroup(db, idParam(req));
    const rows = db
      .prepare(
        `SELECT u.id AS user_id, u.name,
           COALESCE((SELECT SUM(amount_paise) FROM expenses WHERE group_id = g.group_id AND paid_by = u.id), 0)
         - COALESCE((SELECT SUM(s.share_paise) FROM expense_splits s JOIN expenses e ON e.id = s.expense_id WHERE e.group_id = g.group_id AND s.user_id = u.id), 0)
         + COALESCE((SELECT SUM(amount_paise) FROM settlements WHERE group_id = g.group_id AND from_user = u.id), 0)
         - COALESCE((SELECT SUM(amount_paise) FROM settlements WHERE group_id = g.group_id AND to_user = u.id), 0) AS net_paise
         FROM group_members g JOIN users u ON u.id = g.user_id
         WHERE g.group_id = ? ORDER BY u.id`,
      )
      .all(group.id);
    res.json(rows);
  });

  r.post('/groups/:id/settle', (req, res) => {
    const group = requireMembership(db, idParam(req), req.user.id);
    const body = parseBody(settleBody, req.body);
    if (body.from_user === body.to_user) throw new HttpError(400, 'from_user and to_user must differ');
    if (!isMember(db, group.id, body.from_user) || !isMember(db, group.id, body.to_user)) {
      throw new HttpError(400, 'both users must be group members');
    }
    const created_at = new Date().toISOString();
    const result = db
      .prepare('INSERT INTO settlements (group_id, from_user, to_user, amount_paise, created_at) VALUES (?, ?, ?, ?, ?)')
      .run(group.id, body.from_user, body.to_user, body.amount_paise, created_at);
    res.status(201).json({ id: Number(result.lastInsertRowid), group_id: group.id, ...body, created_at });
  });

  return r;
}
